drivers/wifi/infineon/airoc_whd_hal_spi.c

```diff
@@ -0,0 +1,289 @@
+/*
+ * Copyright (c) 2024 Stephen Boylan (stephen.boylan@beechwoods.com)
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <airoc_wifi.h>
+#include "airoc_whd_hal_common.h"
+
+#include <bus_protocols/whd_bus_spi_protocol.h>
+#include <bus_protocols/whd_spi.h>
+#include <whd_wifi_api.h>
+#include <zephyr/drivers/spi.h>
+
+LOG_MODULE_DECLARE(infineon_airoc_wifi, CONFIG_WIFI_LOG_LEVEL);
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+struct whd_bus_priv {
+	whd_spi_config_t spi_config;
+	whd_spi_t spi_obj;
+};
+
+static whd_init_config_t init_config_default = {.thread_stack_size = CY_WIFI_THREAD_STACK_SIZE,
+						.thread_stack_start = NULL,
+						.thread_priority =
+							(uint32_t)CY_WIFI_THREAD_PRIORITY,
+						.country = CY_WIFI_COUNTRY};
+
+/******************************************************
+ *		Function
+ ******************************************************/
+
+int airoc_wifi_init_primary(const struct device *dev, whd_interface_t *interface,
+				whd_netif_funcs_t *netif_funcs,
+				whd_buffer_funcs_t *buffer_if)
+{
+	struct airoc_wifi_data *data = dev->data;
+	const struct airoc_wifi_config *config = dev->config;
+
+	whd_spi_config_t whd_spi_config = {
+		.is_spi_normal_mode = WHD_FALSE,
+	};
+
+#if DT_INST_NODE_HAS_PROP(0, wifi_host_wake_gpios)
+	whd_oob_config_t oob_config = {
+		.host_oob_pin = (void *)&config->wifi_host_wake_gpio,
+		.dev_gpio_sel = DEFAULT_OOB_PIN,
+		.is_falling_edge =
+			(CY_WIFI_HOST_WAKE_IRQ_EVENT == GPIO_INT_TRIG_LOW) ? WHD_TRUE : WHD_FALSE,
+		.intr_priority = CY_WIFI_OOB_INTR_PRIORITY};
+	whd_spi_config.oob_config = oob_config;
+#endif
+#if defined(SPI_DATA_IRQ_SHARED)
+	data->prev_irq_state = 0;
+#endif
+
+	/* Pull bus select line low before enabling WiFi chip */
+	gpio_pin_configure_dt(&config->bus_select_gpio, GPIO_OUTPUT_INACTIVE);
+	gpio_pin_set_dt(&config->bus_select_gpio, 0);
+
+	if (airoc_wifi_power_on(dev)) {
+		LOG_ERR("airoc_wifi_power_on returns fail");
+		return -ENODEV;
```

Comment:
IMO it makes sense to propagate the error number from `airoc_wifi_power_on`

Reply:
I don't think so.  The errors from `airoc_wifi_power_on` are low-level GPIO errors, which are meaningless to applications trying to use WiFi.  I'd say it makes sense here to wrap those in the (slightly) more meaningful ENODEV.